### ISSUE TITLE
systemd-hostnamed start failure: Failed to set up mount namespacing: /run/systemd/unit-root/sys: Invalid argument

### ISSUE THREAD

**Issue by akarollil** [systemd/systemd#16156]
**systemd version the issue has been seen with**
```
test@test:~$ systemctl --version
systemd 245 (245.4-4ubuntu3)
+PAM +AUDIT +SELINUX +IMA +APPARMOR +SMACK +SYSVINIT +UTMP +LIBCRYPTSETUP +GCRYPT +GNUTLS +ACL +XZ +LZ4 +SECCOMP +BLKID +ELFUTILS +KMOD +IDN2 -IDN +PCRE2 default-hierarchy=hybrid
```

**Used distribution**

Freshly installed Ubuntu 20.04 on a VMware VM

```
test@test:~$ cat /etc/issue
Ubuntu 20.04 LTS \n \l
```
**Expected behaviour you didn't see**
`systemd-hostnamed.service` starting up successfully on boot.

**Unexpected behaviour you saw**
systemd-hostnamed service *occasionally* fails to start on boot on a custom OS (Yocto Warrior based, systemd v241) which has another service mount, and then unmount `/sys/firmware/efi/efivars` at the same time as when `systemd-hostnamed.service` starts. There seems to be a race condition when `systemd-hostnamed` tries to remount `/sys` and its submounts as read-only within it's namespace (when using the `ProtectKernelTunables=yes` directive for the service), when there are umounts happening at the same time.

To reproduce this consistently, mount/umount of `/sys/firmware/efi/efivars` was performed in a loop, while simultaneously starting and stopping `systemd-hostnamed.service` in a loop with systemd v241. This was then reproduced consistently with systemd v245 on Ubuntu 20.04. See details below.

Excerpt of errors seen in the service's journal (with `LogLevel=debug` set in `/etc/systemd/system.conf`) from [ubuntu_20.04_systemd-hostnamed_start_fail.log](https://github.com/systemd/systemd/files/4767734/ubuntu_20.04_systemd-hostnamed_start_fail.log):

```
Jun 11 22:35:51 test systemd[29799]: Not remounting /run/systemd/unit-root/sys/fs/cgroup blacklisted by /run/systemd/unit-root/sys/fs/cgroup, called for /run/systemd/unit-root/sys
Jun 11 22:35:51 test systemd[29799]: Not remounting /run/systemd/unit-root/sys/fs/cgroup/unified blacklisted by /run/systemd/unit-root/sys/fs/cgroup, called for /run/systemd/unit-root/sys
Jun 11 22:35:51 test systemd[29799]: Not remounting /run/systemd/unit-root/sys/fs/cgroup/systemd blacklisted by /run/systemd/unit-root/sys/fs/cgroup, called for /run/systemd/unit-root/sys
Jun 11 22:35:51 test systemd[29799]: Not remounting /run/systemd/unit-root/sys/fs/cgroup/pids blacklisted by /run/systemd/unit-root/sys/fs/cgroup, called for /run/systemd/unit-root/sys
Jun 11 22:35:51 test systemd[29799]: Not remounting /run/systemd/unit-root/sys/fs/cgroup/devices blacklisted by /run/systemd/unit-root/sys/fs/cgroup, called for /run/systemd/unit-root/sys
Jun 11 22:35:51 test systemd[29799]: Not remounting /run/systemd/unit-root/sys/fs/cgroup/cpu,cpuacct blacklisted by /run/systemd/unit-root/sys/fs/cgroup, called for /run/systemd/unit-root/sys
Jun 11 22:35:51 test systemd[29799]: Not remounting /run/systemd/unit-root/sys/fs/cgroup/memory blacklisted by /run/systemd/unit-root/sys/fs/cgroup, called for /run/systemd/unit-root/sys
Jun 11 22:35:51 test systemd[29799]: Not remounting /run/systemd/unit-root/sys/fs/cgroup/rdma blacklisted by /run/systemd/unit-root/sys/fs/cgroup, called for /run/systemd/unit-root/sys
Jun 11 22:35:51 test systemd[29799]: Not remounting /run/systemd/unit-root/sys/fs/cgroup/blkio blacklisted by /run/systemd/unit-root/sys/fs/cgroup, called for /run/systemd/unit-root/sys
Jun 11 22:35:51 test systemd[29799]: Not remounting /run/systemd/unit-root/sys/fs/cgroup/cpuset blacklisted by /run/systemd/unit-root/sys/fs/cgroup, called for /run/systemd/unit-root/sys
Jun 11 22:35:51 test systemd[29799]: Not remounting /run/systemd/unit-root/sys/fs/cgroup/freezer blacklisted by /run/systemd/unit-root/sys/fs/cgroup, called for /run/systemd/unit-root/sys
Jun 11 22:35:51 test systemd[29799]: Not remounting /run/systemd/unit-root/sys/fs/cgroup/net_cls,net_prio blacklisted by /run/systemd/unit-root/sys/fs/cgroup, called for /run/systemd/unit-root/sys
Jun 11 22:35:51 test systemd[29799]: Not remounting /run/systemd/unit-root/sys/fs/cgroup/perf_event blacklisted by /run/systemd/unit-root/sys/fs/cgroup, called for /run/systemd/unit-root/sys
Jun 11 22:35:51 test systemd[29799]: Not remounting /run/systemd/unit-root/sys/fs/cgroup/hugetlb blacklisted by /run/systemd/unit-root/sys/fs/cgroup, called for /run/systemd/unit-root/sys
Jun 11 22:35:51 test systemd[29799]: Not remounting /run/systemd/unit-root/sys/fs/bpf blacklisted by /run/systemd/unit-root/sys/fs/bpf, called for /run/systemd/unit-root/sys
Jun 11 22:35:51 test systemd[29799]: Not remounting /run/systemd/unit-root/sys/kernel/tracing blacklisted by /run/systemd/unit-root/sys/kernel/tracing, called for /run/systemd/unit-root/sys
Jun 11 22:35:51 test systemd[29799]: Not remounting /run/systemd/unit-root/sys/kernel/debug blacklisted by /run/systemd/unit-root/sys/kernel/debug, called for /run/systemd/unit-root/sys
Jun 11 22:35:51 test systemd[29799]: Remounted /run/systemd/unit-root/sys read-only.
Jun 11 22:35:51 test systemd[29799]: Remounted /run/systemd/unit-root/sys/kernel/config read-only.
Jun 11 22:35:51 test systemd[29799]: Remounted /run/systemd/unit-root/sys/fs/fuse/connections read-only.
Jun 11 22:35:51 test systemd[29799]: Failed to re-mount '/run/systemd/unit-root/sys' and its submounts: Invalid argument
Jun 11 22:35:51 test systemd[29799]: systemd-hostnamed.service: Failed to set up mount namespacing: /run/systemd/unit-root/sys: Invalid argument
Jun 11 22:35:51 test systemd[29799]: systemd-hostnamed.service: Failed at step NAMESPACE spawning /lib/systemd/systemd-hostnamed: Invalid argument
Jun 11 22:35:51 test systemd[1]: systemd-hostnamed.service: Failed to read oom_kill field of memory.events cgroup attribute: No such file or directory
Jun 11 22:35:51 test systemd[1]: systemd-hostnamed.service: Child 29799 belongs to systemd-hostnamed.service.
Jun 11 22:35:51 test systemd[1]: systemd-hostnamed.service: Main process exited, code=exited, status=226/NAMESPACE
Jun 11 22:35:51 test systemd[1]: systemd-hostnamed.service: Failed with result 'exit-code'.
Jun 11 22:35:51 test systemd[1]: systemd-hostnamed.service: Service will not restart (restart setting)
Jun 11 22:35:51 test systemd[1]: systemd-hostnamed.service: Changed start -> failed
Jun 11 22:35:51 test systemd[1]: systemd-hostnamed.service: Job 33345 systemd-hostnamed.service/start finished, result=failed
Jun 11 22:35:51 test systemd[1]: Failed to start Hostname Service.
```

**Steps to reproduce the problem**

It should be possible to consistently reproduce this issue within about 10 minutes with the following steps:

In one terminal run:

```
while true; do sudo mount -t efivarfs none /sys/firmware/efi/efivars; sudo umount /sys/firmware/efi/efivars; done
```

In another terminal run this script:

```
#!/bin/bash

set -e

while true; do
    systemctl stop systemd-hostnamed
    systemctl start systemd-hostnamed
    systemctl reset-failed systemd-hostnamed
done
```

The starting of the service should eventually fail.

**Comment by poettering** [comment]
I don't get it? Is this Ubuntu or Yocto? You mention both?

Can you provide an strace output of this? i.e. shortly before triggering the issue run "strace -p 1 -s 500 -f strace.log" in another terminal, then provide the relevant part of the "strace.log" file here.

**Comment by akarollil** [comment]
It was initially seen on systemd v241 on Yocto, but then when I was filing an issue for it, saw the issue filing template mention that issues filed here should be for recent versions (guess people don't really follow that rule). So then I reproduced it on v245 on Ubuntu. 

The triggering of the issue is consistent with the repro steps I mention, but the timing isn't. So the stracing would need to be done continuously, but let me see.

**Comment by akarollil** [comment]
strace log when the remount read-only failure happens: [strace_92514.log](https://github.com/systemd/systemd/files/4777057/strace_92514.log) The log was large, so I have trimmed it to when the failing `systemd-hostnamed` gets spawned till when it dies with the mount failure.

`92514` is the PID of `systemd-hostnamed` instance that fails start. Error in service journal when it fails start:

```
Jun 14 18:54:13 test systemd[92514]: systemd-hostnamed.service: Failed to set up mount namespacing: /run/systemd/unit-root/sys: Invalid argument
Jun 14 18:54:13 test systemd[92514]: systemd-hostnamed.service: Failed at step NAMESPACE spawning /lib/systemd/systemd-hostnamed: Invalid argument
Jun 14 18:54:13 test systemd[1]: systemd-hostnamed.service: Main process exited, code=exited, status=226/NAMESPACE
Jun 14 18:54:13 test systemd[1]: systemd-hostnamed.service: Failed with result 'exit-code'.
Jun 14 18:54:13 test systemd[1]: Failed to start Hostname Service.
```

**Comment by tilpner** [comment]
I'm seeing a similar error on NixOS, monitoring only caught it once so far out of hundreds of timer-triggered invocations. Restarting the service worked just fine, no errors. There were no configuration changes for a few days leading up to this error.

```
Failed to set up mount namespacing: /run/systemd/unit-root/: Invalid argument
Failed at step NAMESPACE spawning /nix/store/qflwajndncf0353kh5k869sc1yxgh4p3-unit-script-piclock-ping-start: Invalid argument
Main process exited, code=exited, status=226/NAMESPACE
Failed with result 'exit-code'.
```

The `sys` suffix is not mentioned in my error message, but I don't know if that means these are separate issues, or if it's a configuration difference between NixOS and Ubuntu.

```
$ systemctl --version
systemd 243 (243)
+PAM +AUDIT -SELINUX +IMA +APPARMOR +SMACK -SYSVINIT +UTMP -LIBCRYPTSETUP +GCRYPT -GNUTLS +ACL +XZ +LZ4 +SECCOMP +BLKID -ELFUTILS +KMOD +IDN2 -IDN +PCRE2 default-hierarchy=hybrid
```


<details>
  <summary><b>Journal export</b> (Click to expand!)</summary>

```
__CURSOR=s=3bc817314b144362a316d7aaccf2a7b7;i=12c4b7bf;b=b464b3fc18bd42db85a72ecd18addb59;m=5d3e38653a;t=5b175860cfeba;x=25e7d66b0d14dddd
__REALTIME_TIMESTAMP=1602493201841850
__MONOTONIC_TIMESTAMP=400475841850
_BOOT_ID=b464b3fc18bd42db85a72ecd18addb59
_TRANSPORT=journal
_SELINUX_CONTEXT=kernel
_SYSTEMD_SLICE=system.slice
_MACHINE_ID=8e08b6696a304ca99f5196d306860df1
SYSLOG_FACILITY=3
SYSLOG_IDENTIFIER=systemd
_UID=0
_GID=0
_CAP_EFFECTIVE=3fffffffff
PRIORITY=3
UNIT=piclock-ping.service
INVOCATION_ID=a5b226a1d09e4424aab4ad365c8ed168
CODE_FILE=src/core/execute.c
CODE_LINE=3398
CODE_FUNC=exec_child
MESSAGE=piclock-ping.service: Failed to set up mount namespacing: /run/systemd/unit-root/: Invalid argument
_PID=32503
_COMM=(ng-start)
_SYSTEMD_CGROUP=/system.slice/piclock-ping.service
_SYSTEMD_UNIT=piclock-ping.service
_SYSTEMD_INVOCATION_ID=a5b226a1d09e4424aab4ad365c8ed168
_SOURCE_REALTIME_TIMESTAMP=1602493201841843

__CURSOR=s=3bc817314b144362a316d7aaccf2a7b7;i=12c4b7c0;b=b464b3fc18bd42db85a72ecd18addb59;m=5d3e386677;t=5b175860cfff7;x=baa2f56b8695b2d8
__REALTIME_TIMESTAMP=1602493201842167
__MONOTONIC_TIMESTAMP=400475842167
_BOOT_ID=b464b3fc18bd42db85a72ecd18addb59
_TRANSPORT=journal
_SELINUX_CONTEXT=kernel
_SYSTEMD_SLICE=system.slice
_MACHINE_ID=8e08b6696a304ca99f5196d306860df1
SYSLOG_FACILITY=3
SYSLOG_IDENTIFIER=systemd
_UID=0
_GID=0
_CAP_EFFECTIVE=3fffffffff
PRIORITY=3
UNIT=piclock-ping.service
INVOCATION_ID=a5b226a1d09e4424aab4ad365c8ed168
CODE_FILE=src/core/execute.c
_PID=32503
_COMM=(ng-start)
_SYSTEMD_CGROUP=/system.slice/piclock-ping.service
_SYSTEMD_UNIT=piclock-ping.service
_SYSTEMD_INVOCATION_ID=a5b226a1d09e4424aab4ad365c8ed168
CODE_LINE=3894
CODE_FUNC=exec_spawn
MESSAGE_ID=641257651c1b4ec9a8624d7a40a9e1e7
MESSAGE=piclock-ping.service: Failed at step NAMESPACE spawning /nix/store/qflwajndncf0353kh5k869sc1yxgh4p3-unit-script-piclock-ping-start: Invalid argument
EXECUTABLE=/nix/store/qflwajndncf0353kh5k869sc1yxgh4p3-unit-script-piclock-ping-start
_SOURCE_REALTIME_TIMESTAMP=1602493201841875

__CURSOR=s=3bc817314b144362a316d7aaccf2a7b7;i=12c4b7c4;b=b464b3fc18bd42db85a72ecd18addb59;m=5d3e386938;t=5b175860d02b8;x=e73f69d8142bde7a
__REALTIME_TIMESTAMP=1602493201842872
__MONOTONIC_TIMESTAMP=400475842872
_BOOT_ID=b464b3fc18bd42db85a72ecd18addb59
_TRANSPORT=journal
_SELINUX_CONTEXT=kernel
_MACHINE_ID=8e08b6696a304ca99f5196d306860df1
SYSLOG_FACILITY=3
SYSLOG_IDENTIFIER=systemd
_PID=1
_UID=0
_GID=0
_COMM=systemd
_EXE=/nix/store/grsj02sxnkgsa1hgl9rnhcwnd5p96gi7-systemd-243.7/lib/systemd/systemd
_CMDLINE=systemd
_CAP_EFFECTIVE=3fffffffff
_SYSTEMD_CGROUP=/init.scope
_SYSTEMD_UNIT=init.scope
_SYSTEMD_SLICE=-.slice
PRIORITY=5
CODE_FILE=src/core/unit.c
UNIT=piclock-ping.service
INVOCATION_ID=a5b226a1d09e4424aab4ad365c8ed168
CODE_LINE=5800
CODE_FUNC=unit_log_process_exit
MESSAGE_ID=98e322203f7a4ed290d09fe03c09fe15
MESSAGE=piclock-ping.service: Main process exited, code=exited, status=226/NAMESPACE
EXIT_CODE=exited
EXIT_STATUS=226
COMMAND=ExecStart
_SOURCE_REALTIME_TIMESTAMP=1602493201842543

__CURSOR=s=3bc817314b144362a316d7aaccf2a7b7;i=12c4b7c5;b=b464b3fc18bd42db85a72ecd18addb59;m=5d3e386975;t=5b175860d02f5;x=56356c1195823688
__REALTIME_TIMESTAMP=1602493201842933
__MONOTONIC_TIMESTAMP=400475842933
_BOOT_ID=b464b3fc18bd42db85a72ecd18addb59
_TRANSPORT=journal
_SELINUX_CONTEXT=kernel
_MACHINE_ID=8e08b6696a304ca99f5196d306860df1
SYSLOG_FACILITY=3
PRIORITY=4
SYSLOG_IDENTIFIER=systemd
_PID=1
_UID=0
_GID=0
_COMM=systemd
_EXE=/nix/store/grsj02sxnkgsa1hgl9rnhcwnd5p96gi7-systemd-243.7/lib/systemd/systemd
_CMDLINE=systemd
_CAP_EFFECTIVE=3fffffffff
_SYSTEMD_CGROUP=/init.scope
_SYSTEMD_UNIT=init.scope
_SYSTEMD_SLICE=-.slice
CODE_FILE=src/core/unit.c
UNIT=piclock-ping.service
INVOCATION_ID=a5b226a1d09e4424aab4ad365c8ed168
CODE_LINE=5756
CODE_FUNC=unit_log_failure
MESSAGE_ID=d9b373ed55a64feb8242e02dbe79a49c
MESSAGE=piclock-ping.service: Failed with result 'exit-code'.
UNIT_RESULT=exit-code
_SOURCE_REALTIME_TIMESTAMP=1602493201842612
```
</details>

**Comment by real-artswan** [comment]
Very similar to @tilpner on Debian testing
```
$ lsb_release -a
No LSB modules are available.
Distributor ID: Debian
Description:    Debian GNU/Linux bullseye/sid
Release:        testing
Codename:       bullseye
```

```
$ systemctl --version
systemd 247 (247.1-3)
+PAM +AUDIT +SELINUX +IMA +APPARMOR +SMACK +SYSVINIT +UTMP +LIBCRYPTSETUP +GCRYPT +GNUTLS +ACL +XZ +LZ4 +ZSTD +SECCOMP +BLKID +ELFUTILS +KMOD +IDN2 -IDN +PCRE2 default-hierarchy=hybrid
```

```
Failed to set up mount namespacing: /run/systemd/unit-root/: Input/output error
Failed at step NAMESPACE spawning /lib/systemd/systemd-hostnamed: Input/output error
Main process exited, code=exited, status=226/NAMESPACE
Failed with result 'exit-code'.
```

The only difference I see that it says "Input/output error" instead of "Invalid argument"
Restarting works just fine without errors.

**Comment by Harvie** [comment]
I get same issue on ArchLinux. This is so annoying. Is there any workaround yet? Downgrading is not easy in my case.

**Comment by polrus** [comment]
I think the same issue in Opensuse tumbleweed
https://bugzilla.suse.com/show_bug.cgi?id=1181681

**Comment by poettering** [comment]
@polrus your case is different, you see ENOENT when the logic tries to remount /etc/named.conf.include. what kind of mount is that?

**Comment by polrus** [comment]
I have no idea - but solved it by removing all packages related to DNS and reinstallig the packages. 
Now I can see the file exists but it's empty
-rw-r--r-- 1 root  named    0 Feb 11 10:54 /etc/named.conf.include
and systemd does not complain anymore

**Comment by Harvie** [comment]
I've fixed the issue by adding `_netdev` option to my SSHFS mount in fstab. This is annoying.
I have specified `delay_connect` sshfs option in fstab. This worked very well for years. Now it is broken.
It basicaly says that SSHFS should act as if mount was succesful and then later mount it once the remote server becomes available. Now the systemd is preventing that behaviour in some way. giving `_netdev` flag does postpone sshfs mount after network connection, but i guess when i will be connected to the internet, but the remote server would be down it will still give me a headache.

Is there some flag to say that i want to disable all this smart-ass stuff for this mountpoint and just mount it the old way?

**Comment by morfikov** [comment]
I also have a similar issue. This one ends with `Host is down` (and also affects more services, like upower):

```
systemd[1]: Starting systemd-hostnamed.service...
systemd[288098]: systemd-hostnamed.service: Failed to set up mount namespacing: /run/systemd/unit-root/: Host is down
systemd[288098]: systemd-hostnamed.service: Failed at step NAMESPACE spawning /lib/systemd/systemd-hostnamed: Host is down
systemd[1]: systemd-hostnamed.service: Main process exited, code=exited, status=226/NAMESPACE
systemd[1]: systemd-hostnamed.service: Failed with result 'exit-code'.
systemd[1]: Failed to start systemd-hostnamed.service.
```
It looks the the service has issues when some remote SMB share is mounted on the host, and the remote host disappears (is turned off) while the share is still mounted. I have the following line in the /etc/fstab file:
`
//192.168.1.239/data /media/rpi-tv cifs guest,users,noauto,nofail,vers=3.11 0 0`

When I manually unmount the SMB share, then the systemd-hostnamed.service can be restarted without any problems. 

```
# systemd --version
systemd 247 (247.3-3)
+PAM +AUDIT +SELINUX +IMA +APPARMOR +SMACK +SYSVINIT +UTMP +LIBCRYPTSETUP +GCRYPT +GNUTLS +ACL +XZ +LZ4 +ZSTD +SECCOMP +BLKID +ELFUTILS +KMOD +IDN2 -IDN +PCRE2 default-hierarchy=unified

# cat /proc/version
Linux version 5.11.5-amd64 (morfik@morfikownia) (gcc (Debian 10.2.1-23) 10.2.1 20210312, GNU ld (GNU Binutils for Debian) 2.35.2) #9 SMP PREEMPT Sat Mar 13 17:48:02 CET 2021
```

**Comment by Harvie** [comment]
Having systemd refuse to boot properly because some other system in network didn't managed to boot properly is really stupid and it will eventualy lead to gridlock chain reaction shuting down the whole internet. :-)

**Comment by poettering** [comment]
Any chance you can test #19101? It should fix this issue?

**Comment by RobbieTheK** [comment]
This problem is happening again I added a [strace at Bugzilla for Fedora](https://bugzilla.redhat.com/show_bug.cgi?id=1885090#c15).

**Comment by xundeenergie** [comment]
Debian Bullseye too...

**Comment by elboulangero** [comment]
I had a very similar issue:

```
Jun 29 00:00:02 HOST systemd[3301718]: logrotate.service: Failed to set up mount namespacing: /run/systemd/unit-root/dev: Invalid argument
Jun 29 00:00:02 HOST systemd[3301718]: logrotate.service: Failed at step NAMESPACE spawning /usr/sbin/logrotate: Invalid argument
Jun 29 00:00:02 HOST systemd[1]: logrotate.service: Main process exited, code=exited, status=226/NAMESPACE
Jun 29 00:00:02 HOST systemd[1]: logrotate.service: Failed with result 'exit-code'.
```

I could troubleshoot it just by running `journalctl -f` in another terminal:

```
Jun 29 01:28:52 HOST systemd[1]: /dev/null is not a device.
```

Indeed:

```
  # ls -l /dev/null
  -rw-rw--w- 1 root root 1257 Jun 29 02:19 /dev/null
```

Why would `/dev/null` not be a device? In my case, this was due to a bug in logrotate. Running `logrotate -s /dev/null ...` replaces `/dev/null` with a regular file, thus breaking the system... [This was fixed upstream in version 3.19](https://github.com/logrotate/logrotate/blob/e9277f77af70854ed06b4a682d7608b6586fabf9/ChangeLog.md?plain=1#L46)

So in short: not a bug in systemd, and simply looking at the journal was enough to understand the issue.

**Comment by papadakospan** [comment]
Thank you @elboulangero ! Your comment helped me track it down to a systemd service I'd made.

When I saw your comment my mind went immediately to an "experiment" I had done.

Suggestion for people who see this: Use grep to find the culprit.

`grep -R '/dev/null' /etc/systemd/system/`
